Provides the same json files as product details from our database

sylvestre opened the pull request:
As part of my Q1 & Q2, I would like to kill product detail. This will also remove one of the last blocker to remove SVN mozilla.

This patch introduces a new field the database to store the ship date ( https://bugzilla.mozilla.org/show_bug.cgi?id=1101596 )

It also provides the same files as in product details:
https://bugzilla.mozilla.org/show_bug.cgi?id=1083718

And, last but not least, I wrote an importer to update the shipped date to the correct one ( https://bugzilla.mozilla.org/show_bug.cgi?id=1149978 )

There is a thing to fix with ESR 31 and 38 but I would like us to deploy that asap to work with IT about the publication of the json files.

sylvestre commented:
I don't mind splitting the patch if you want.

rail commented on kickoff/importjson.py:
```diff
@@ -0,0 +1,12 @@
+import json
+json_data=open('/home/sylvestre/dev/mozilla/product-details.svn/json/languages.json')
```
No such file or directory :P

bhearsum commented on kickoff/fennecversions.py:
```diff
@@ -0,0 +1,5 @@
+fennecversions = {
```
This file isn't used by anything AFAICT. Is that a bug? If not, do we even need this?

sylvestre replied:
Right, this is useless.

bhearsum commented on kickoff/firefoxdetails.py:
```diff
@@ -0,0 +1,784 @@
+from firefoxversions import firefoxversions as versions
+
+primary_builds = {
```
I feel like this (and some other) data structures could be done in a better way, but I'm having a lot of trouble understanding the overall structure. I think it would help a lot if you could explain a couple of things:
1) What data are we currently missing from Ship It that we need to generate the JSON files.
2) Which data (existing and new) goes with which endpoints?

sylvestre replied:
We need a few things:
* the shipped date
* the current aurora version (we don't have this information in ship-it yet)
* a way to differentiate ESR & ESR_NEXT (currently 31 & 38)
* list of supported languages
* for all the locales, which versions (aurora, beta, release) are supported

About the data which are used by endpoints, it is mainly version name + shipped date.

sylvestre commented:
I know this is not perfect but this is a first version. I am planning to iterate on this once we moved away from p-d & svn.

bhearsum commented on migrate_repo/versions/008_Add_shipped_at.py:
```diff
@@ -0,0 +1,25 @@
+        shippedAt.create(table)
+    add_shippedAt(Table('fennec_release', metadata, autoload=True))
+    add_shippedAt(Table('firefox_release', metadata, autoload=True))
+    add_shippedAt(Table('thunderbird_release', metadata, autoload=True))
```
What's your plan/hope for filling in historical data accurately? This upgrade method is going to end up with every existing release having an inaccurate shipped date, obviously.

sylvestre replied:
This is implemented here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1149978
(a script which will extract from the json info the correct date)

sylvestre commented:
Discussing with Ben on vidyo, we need to:
* generate the beta + esr + release locales from the ship-it database (kickoff/firefoxdetails.py)
* aurora & nightly will be hardcoded as we don't have this information in ship-it
* try to guess the ESR & ESR_NEXT version
* aurora & nightly will probably have to be hardcoded as it is a human task
* use https://github.com/mozilla/ship-it/blob/master/vendor/lib/python/mozilla/build/versions.py#L6 for getFilteredReleases
* see also https://github.com/mozilla/balrog/blob/master/auslib/util/versions.py#L5

sylvestre commented:
I implemented all we discussed except the ESR vs ESR_NEXT detection

sylvestre commented:
New changes:
* I removed the beta build, I am pretty sure this is not used. The current file is empty: http://viewvc.svn.mozilla.org/vc/libs/product-details/json/firefox_beta_builds.json?revision=141468&view=markup
* I hardcoded the ESR & ESR_NEXT release. I am not sure how to compute that correctly and we are doing that every 8 months. So, we can still do that manually (and I can come back on this later)
* I added some information about the changes on the landing page: http://127.0.0.1:5000/json_exports.html

To me, it is ready to be merged now

bhearsum commented:
I should get a chance to look at this tomorrow - sorry I haven't been able to yet.

On Tue, Jun 16, 2015 at 07:47:24AM -0700, Sylvestre Ledru wrote:
> New changes:
> * I removed the beta build, I am pretty sure this is not used. The current file is empty: http://viewvc.svn.mozilla.org/vc/libs/product-details/json/firefox_beta_builds.json?revision=141468&view=markup
> * I hardcoded the ESR & ESR_NEXT release. I am not sure how to compute that correctly and we are doing that every 8 months. So, we can still do that manually (and I can come back on this later)
> * I added some information about the changes on the landing page: http://127.0.0.1:5000/json_exports.html
> 
> To me, it is ready to be merged now
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/mozilla/ship-it/pull/23#issuecomment-112456469

sylvestre commented:
I added tests for ESR & ESR_NEXT and fixed a few things.

bhearsum commented on requirements/prod.txt:
```diff
@@ -5,3 +5,4 @@ flask-wtf==0.8
 flask-sqlalchemy==0.16
 sqlalchemy-migrate==0.7.2
 cef==0.5
+simplejson
```
This belongs in compiled.txt, because our webheads have the C version installed on them. prod.txt is for things that go in the embedded vendor library.

sylvestre replied:
fixed

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+import config
+
+
+def getFilteredReleases(product, categories, ESR_NEXT=False, lastRelease=None, withL10N=False):
```
This is still using custom regexes rather than ANY_VERSION_REGEX. I'm not r-'ing over this necessarily, but what's the rationale?

sylvestre replied:
I forgot :(

sylvestre replied:
OK, I had a look and we can't use it because we have too many special cases. Some examples: 3.6.0build1, 3.0rc1 or 3.6.3plugin1 are not understood by this regexp.

bhearsum commented on migrate_repo/versions/008_Add_shipped_at.py:
```diff
@@ -0,0 +1,25 @@
+
+    def add_shippedAt(table):
+        shippedAt = Column('shippedAt', DateTime(pytz.utc))
+        shippedAt.create(table)
```
What are we going to do about backfilling old release ship dates? Do we need to?

sylvestre replied:
This is https://bugzilla.mozilla.org/show_bug.cgi?id=1149978 => already done.

bhearsum replied:
OK, we'll have to do that after this patch lands, since it depends on "shippedAt" existing. No big deal though, thanks for preparing the SQL!

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+def firefox_history_major_releases_json():
+    # Match X.Y and 14.0.1 (special case)
+    values = getFilteredReleases("firefox", "major")
+    return jsonify(values)
```
Most of these methods are looking good, great job!

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+def firefox_versions_json():
+    # Stable
+    lastStable = getFilteredReleases("firefox", ["major", "stability"], lastRelease=True)
+    firefoxversions["LATEST_FIREFOX_VERSION"] = lastStable[0][0]
```
Overriding a global data structure like firefoxversions is probably going to cause issues at some point, especially if the app ever runs in multiple threads. There's a couple of ways you can go about fixing this:
1) Get rid of firefoxversions.py (and similar files) and replace it with a Jinja template that contains the substitution logic.
2) Make a copy of firefoxosversions in this function.

The first option is probably more flexible, and you'll need to fix this for the mobile and thunderbird parts too.

bhearsum replied:
Am I reading the new diff correctly that you took a new approach and eliminate the need for a file like firefoxversions.py altogether? That's totally fine, just double checking!

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+            i += 1
+
+
+def generateLocateBuilds(buildsVersionLocales, l10nchangesets, lastVersion):
```
Naming nit: locate vs. localized. (locate is a verb)

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+    # The l10n format for desktop is 'locale changeset'
+    # parse it
+    for locale in l10nchangesets.split('\n'):
+        if "\t" in locale:
```
We've got an existing parser for this :). https://github.com/mozilla/build-tools/blob/master/lib/python/release/l10n.py#L39

Please use it unless there's a reason you can't.

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+    return buildsVersionLocales
+
+
+def updateBuilds(product):
```
Can you find a better name for this method? It's not really clear to me what it's doing.

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+    lastStable = getFilteredReleases(product, ["esr"], lastRelease=True, withL10N=True)
+    buildsVersionLocales = generateLocateBuilds(buildsVersionLocales, lastStable[0][2],
+                                                lastStable[0][0])
+    buildsVersionLocales = fillAuroraVersion(buildsVersionLocales)
```
All this reassignment of buildsVersionLocales is confusing...this would be more readable if generateLocateBuilds and fillAuroraVersion returned new dicts, and you called .update() on buildsVersionLocales instead. Eg:
buildsVersionLocales = {}
buildsVersionLocales.update(generateLocateBuilds(lastStable[0][2], lastStable[0][0]))

I _think_ you'll end up with the same result since you're not using deeply nested dicts.

sylvestre replied:
This does not work. update() replaces the value, not add them 
```


dict = {'Name': 'Zara', 'Age': 7}
dict2 = {'Age': '8' }

dict.update(dict2)
print "Value : %s" %  dict
```
=> Value : {'Age': '8', 'Name': 'Zara'}

bhearsum commented on kickoff/jsonexport.py:
```diff
@@ -0,0 +1,244 @@
+# @app.route('/json/firefox_beta_builds.json', methods=['GET'])
+# def firefox_beta_builds_json():
+#     updateBuilds("firefox")
+#     return jsonify(beta_builds)
```
Let's just get rid of this if it's unused.

bhearsum commented on kickoff/model.py:
```diff
@@ -18,6 +19,8 @@ class Release(object):
     _submittedAt = db.Column('submittedAt', db.DateTime(pytz.utc),
                              nullable=False, default=datetime.utcnow)
+    _shippedAt = db.Column('shippedAt', db.DateTime(pytz.utc),
+                             nullable=False, default=datetime.utcnow)
```
Are you sure defaulting this to utcnow will work? I think this means that shippedAt will end up getting set to the current utc datetime when the release is _submitted_, which I don't think you want. I think you need this to default to NULL and update it when the "Shipped!" button is pressed.

sylvestre commented:
Thanks, I will address the comments tomorrow.

bhearsum commented on config.py:
```diff
@@ -0,0 +1,7 @@
+# Not used
```
Missed this in the earlier round -- I don't think this is going to work for the WSGI app, only modules under "kickoff" will be importable, can you move this to something inside of that directory?

bhearsum commented:
I think this is ready to go once the config.py file gets moved and I make sure I understand what happened to firefoxversions.py and co.

sylvestre commented:
I followed your advice for firefoxversions.py (template) & moved config.py.

bhearsum commented:
OK, I'm satisfied with this.

lizzard commented:
could this have broken ship-it?  https://bugzilla.mozilla.org/show_bug.cgi?id=1176759

bhearsum commented:
Yep!

On Mon, Jun 22, 2015 at 02:48:45PM -0700, Liz wrote:
> could this have broken ship-it?  https://bugzilla.mozilla.org/show_bug.cgi?id=1176759 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/mozilla/ship-it/pull/23#issuecomment-114273886
